src/test/kotlin/com/terraformation/backend/funder/FundingEntityServiceTest.kt

```diff
@@ -0,0 +1,263 @@
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+import org.springframework.security.access.AccessDeniedException
+
+class FundingEntityServiceTest : DatabaseTest(), RunsAsUser {
+  override val user: TerrawareUser = mockUser()
+
+  private val clock = TestClock()
+  private lateinit var fundingEntityStore: FundingEntityStore
+  private lateinit var fundingEntityUserStore: FundingEntityUserStore
+  private lateinit var fundingEntityId: FundingEntityId
+  private lateinit var service: FundingEntityService
+
+  @BeforeEach
+  fun setUp() {
+    fundingEntityStore = FundingEntityStore(dslContext)
+    fundingEntityUserStore = FundingEntityUserStore(dslContext)
+    service = FundingEntityService(clock, dslContext, fundingEntityStore, fundingEntityUserStore)
+
+    fundingEntityId = insertFundingEntity()
+    insertOrganization()
+
+    every { user.canReadFundingEntities() } returns true
+    every { user.canCreateFundingEntities() } returns true
+    every { user.canDeleteFundingEntities() } returns true
+    every { user.canUpdateFundingEntities() } returns true
+    every { user.canUpdateFundingEntityProjects() } returns true
+  }
+
+  @Test
+  fun `create requires user to be able to manage funding entities`() {
+    every { user.canCreateFundingEntities() } returns false
+
+    assertThrows<AccessDeniedException> { service.create("Some Other Entity") }
+  }
+
+  @Test
+  fun `create populates created and modified fields`() {
+    val newTime = clock.instant().plusSeconds(1000)
+    clock.instant = newTime
+
+    val newUserId = insertUser()
+
+    val name = "New Funding Entity"
+    val row = FundingEntitiesRow(name = name)
+
+    every { user.userId } returns newUserId
+
+    val createdModel = service.create(row.name!!)
+
+    val expected =
+        row.copy(
+            createdBy = user.userId,
+            createdTime = clock.instant(),
+            id = createdModel.id,
+            modifiedBy = newUserId,
+            modifiedTime = clock.instant(),
+            name = name,
+        )
+    // retrieving both row and model to check createdBy/modifiedBy as well as projects
+    val actualRow = fundingEntitiesDao.fetchOneById(createdModel.id)!!
+    val actualModel = fundingEntityStore.fetchOneById(createdModel.id)
+
+    assertEquals(expected, actualRow)
+    assertEquals(0, actualModel.projects!!.size)
+  }
+
+  @Test
+  fun `create rejects duplicates by name`() {
+    assertThrows<FundingEntityExistsException> { service.create("TestFundingEntity") }
+  }
+
+  @Test
+  fun `create adds project rows`() {
+    insertOrganization()
+    val projectId1 = insertProject()
+    val projectId2 = insertProject()
+    val projectSet = setOf(projectId1, projectId2)
+
+    val inserted = service.create("New Funding Entity with Projects", projectSet)
+
+    val actual = fundingEntityStore.fetchOneById(inserted.id)
+
+    assertEquals(2, actual.projects!!.size)
+    assertEquals(projectSet, actual.projects!!.toSet())
```

Comment:
The size assertion isn't needed here; the set comparison will fail if the sizes aren't equal, and the assertion failure message will be more useful in that case since you'll be able to see which specific values were missing or added.